<!DOCTYPE html>
<!-- saved from url=(0014)about:internet -->
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
<meta http-equiv="x-ua-compatible" content="IE=9" >

<title>Understanding Natural Selection from Population Genomic Data</title>

<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 12px;
   margin: 8px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 { 
   font-size:2.2em; 
}

h2 { 
   font-size:1.8em; 
}

h3 { 
   font-size:1.4em; 
}

h4 { 
   font-size:1.0em; 
}

h5 { 
   font-size:0.9em; 
}

h6 { 
   font-size:0.8em; 
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre {	
   margin-top: 0;
   max-width: 95%;
   border: 1px solid #ccc;
   white-space: pre-wrap;
}

pre code {
   display: block; padding: 0.5em;
}

code.r, code.cpp {
   background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * { 
      background: transparent !important; 
      color: black !important; 
      filter:none !important; 
      -ms-filter: none !important; 
   }

   body { 
      font-size:12pt; 
      max-width:100%; 
   }
       
   a, a:visited { 
      text-decoration: underline; 
   }

   hr { 
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote { 
      padding-right: 1em; 
      page-break-inside: avoid; 
   }

   tr, img { 
      page-break-inside: avoid; 
   }

   img { 
      max-width: 100% !important; 
   }

   @page :left { 
      margin: 15mm 20mm 15mm 10mm; 
   }
     
   @page :right { 
      margin: 15mm 10mm 15mm 20mm; 
   }

   p, h2, h3 { 
      orphans: 3; widows: 3; 
   }

   h2, h3 { 
      page-break-after: avoid; 
   }
}

</style>





</head>

<body>
<h1>Understanding Natural Selection from Population Genomic Data</h1>

<p>still in progress (not tested). Using Duffy as an example <a href="http://www.snpedia.com/index.php/Rs2814778">http://www.snpedia.com/index.php/Rs2814778</a></p>

<h2>1. Setting up your Enviroment</h2>

<h3>1.1 Downloading and Building Methods</h3>

<h3>1.2 Downloading Data</h3>

<h4>1000genomes_phase3</h4>

<pre><code class="bash">cd data/

# download and filter a region (1:145000000-175000000) from the 1000genomes project
tabix -h ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr1.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz 1:145000000-175000000 | gzip -c &gt; 1000genomes_phase3_chr1_145000000_175000000.vcf.gz

# download 1000genomes_phase3 individual information
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/integrated_call_samples_v3.20130502.ALL.panel
</code></pre>

<h3>1.3 Filtering Data and QC</h3>

<h4>Preparing CLST Files</h4>

<pre><code class="bash"># remove the first line and format file as IID | IID | CLST for YRI, CEU, and CHB
sed &#39;1d&#39; integrated_call_samples_v3.20130502.ALL.panel | awk &#39;{if($3 == &quot;YRI&quot; || $3 == &quot;CEU&quot; || $3 == &quot;CHB&quot;) print $1, $1, $3} &gt; yri_ceu_chb_clst.txt

# write a file of just IIDs 
awk &#39;{print $1} yri_ceu_chb_clst.txt &gt; yri_ceu_chb_iids.txt

# write a file of just IIDs from YRI 
awk &#39;{if($3 == &quot;YRI&quot;) print $1}&#39; &gt; yri_iids.txt

cd ..
</code></pre>

<h4>Preparing Genomic Data</h4>

<pre><code class="bash"># filter out indels, multi-allelic sites, and individuals not from YRI, CEU, CHB 
# this will be used for cross-population based selection analysis
vcftools --gzvcf data/1000genomes_phase3_chr1_145000000_175000000.vcf.gz \
         --remove-indels \
         --keep data/yri_ceu_chb_iids.txt \
         --recode \
         --out data/yri_ceu_chb_chr1_145000000_175000000_filtered

# filter out indivduals not from YRI
# this will be used for haplotype based selection analysis 
vcftools --gzvcf data/yri_ceu_chb_chr1_145000000_175000000_filtered.vcf.gz \
         --keep data/yri_iids.txt  \
         --recode \
         --out data/yri_chr1_145000000_175000000_filtered
</code></pre>

<h3>Questions</h3>

<h2>2. Computing Selection Statistics</h2>

<h3>2.1 iHS</h3>

<h3>2.2 nSL</h3>

<h3>2.3. Fst</h3>

<h3>2.4 PBS</h3>

<h3>2.5 Visualization</h3>

<h3>Questions</h3>

<h2>3. Assesing Significance</h2>

<h3>3.1 Genome(Region)-Wide Null Distribution</h3>

<h3>3.2 Simulating a Null Model</h3>

<h3>3.3 Visualization</h3>

<h3>3.4 Spandrels of San Marco (Caveats)</h3>

<h3>Questions</h3>

</body>

</html>

