<!DOCTYPE html>
<!-- saved from url=(0014)about:internet -->
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
<meta http-equiv="x-ua-compatible" content="IE=9" >

<title>Understanding Natural Selection from Population Genomic Data</title>

<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 12px;
   margin: 8px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 { 
   font-size:2.2em; 
}

h2 { 
   font-size:1.8em; 
}

h3 { 
   font-size:1.4em; 
}

h4 { 
   font-size:1.0em; 
}

h5 { 
   font-size:0.9em; 
}

h6 { 
   font-size:0.8em; 
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre {	
   margin-top: 0;
   max-width: 95%;
   border: 1px solid #ccc;
   white-space: pre-wrap;
}

pre code {
   display: block; padding: 0.5em;
}

code.r, code.cpp {
   background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * { 
      background: transparent !important; 
      color: black !important; 
      filter:none !important; 
      -ms-filter: none !important; 
   }

   body { 
      font-size:12pt; 
      max-width:100%; 
   }
       
   a, a:visited { 
      text-decoration: underline; 
   }

   hr { 
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote { 
      padding-right: 1em; 
      page-break-inside: avoid; 
   }

   tr, img { 
      page-break-inside: avoid; 
   }

   img { 
      max-width: 100% !important; 
   }

   @page :left { 
      margin: 15mm 20mm 15mm 10mm; 
   }
     
   @page :right { 
      margin: 15mm 10mm 15mm 20mm; 
   }

   p, h2, h3 { 
      orphans: 3; widows: 3; 
   }

   h2, h3 { 
      page-break-after: avoid; 
   }
}

</style>



<!-- MathJax scripts -->
<script type="text/javascript" src="https://c328740.ssl.cf1.rackcdn.com/mathjax/2.0-latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
</script>



</head>

<body>
<h1>Understanding Natural Selection from Population Genomic Data</h1>

<ul>
<li>still in progress (not tested). </li>
<li>using Duffy as an example (<a href="http://www.snpedia.com/index.php/Rs2814778">http://www.snpedia.com/index.php/Rs2814778</a>)</li>
</ul>

<h2>1. Setting up your Enviroment</h2>

<h3>1.1 Downloading and Building Methods</h3>

<h4>Download</h4>

<pre><code class="bash">cd src/

# download tabix
git clone https://github.com/samtools/tabix

# download htslib &amp; bcftools
wget http://sourceforge.net/projects/samtools/files/samtools/1.2/htslib-1.2.1.tar.bz2
wget http://sourceforge.net/projects/samtools/files/samtools/1.2/bcftools-1.2.tar.bz2

# clone selscan
git clone https://github.com/szpiech/selscan

# download plink (double check that the correct OS is specified)
wget https://www.cog-genomics.org/static/bin/plink150602/plink_linux_x86_64_dev.zip
</code></pre>

<h4>Build</h4>

<pre><code class="bash"># tabix
cd tabix/
make
mv tabix ../../bin/

# bcftools
cd ../
tar xvjf htslib-1.2.1.tar.bz2
tar xvjf bcftools-1.2.tar.bz2
cd bcftools/
make
mv bcftools ../../bin/

# plink
cd ../
mkdir plink
unzip plink_linux_x86_64_dev.zip -d plink_linux_x86_64_dev.zip
mv plink/plink ../bin

# selscan
cd selscan/

# open Makefile make sure the correct OS is specified 
cd src/ 
make
mv norm ../../bin/
mv selscan ../../bin/

# lets go home
cd ../../
ls bin/

# you should see binaries for tabix, bcftools, plink, selscan, and norm
</code></pre>

<h3>1.2 Downloading Data</h3>

<h4>1000genomes_phase3</h4>

<pre><code class="bash">cd data/

# download and filter a region (1:150000000-170000000) from the 1000genomes project
../bin/tabix -h ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr1.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz 1:150000000-170000000 | gzip -c &gt; 1000genomes_phase3_chr1_150000000_170000000.vcf.gz

# download 1000genomes_phase3 individual information
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/integrated_call_samples_v3.20130502.ALL.panel
</code></pre>

<h3>1.3 Filtering Data and QC</h3>

<h4>Preparing CLST / IID Files</h4>

<pre><code class="bash"># remove the first line and format file as IID | IID | CLST for YRI, CEU, and CHB
sed &#39;1d&#39; integrated_call_samples_v3.20130502.ALL.panel | awk &#39;{if($2 == &quot;YRI&quot; || $2 == &quot;CEU&quot; || $2 == &quot;CHB&quot;) print $1, $1, $2}&#39; &gt; yri_ceu_chb_clst.txt

# write a file of just IIDs 
awk &#39;{print $1}&#39; yri_ceu_chb_clst.txt &gt; yri_ceu_chb_iids.txt

# write a clst file for just YRI and CEU
awk &#39;{if($3 == &quot;YRI&quot; || $3 == &quot;CEU&quot;) print $0}&#39; yri_ceu_chb_clst.txt &gt; yri_ceu_clst.txt

# write a clst file for just YRI and CHB 
awk &#39;{if($3 == &quot;YRI&quot; || $3 == &quot;CHB&quot;) print $0}&#39; yri_ceu_chb_clst.txt &gt; yri_chb_clst.txt

# write a clst file for just CEU and CHB
awk &#39;{if($3 == &quot;CEU&quot; || $3 == &quot;CHB&quot;) print $0}&#39; yri_ceu_chb_clst.txt &gt; ceu_chb_clst.txt


# write a file of just IIDs from CHB 
awk &#39;{if($3 == &quot;CHB&quot;) print $1}&#39; yri_ceu_chb_clst.txt &gt; chb_iids.txt

# check if # individuals are correct
wc -l ceu_iids.txt chb_iids.txt yri_iids.txt
</code></pre>

<h4>Preparing Genomic Data</h4>

<pre><code class="bash"># filter out indels (-I), multi-allelic sites (-c 2 -C 2), and keep individuals (-S) from YRI, CEU, CHB 
# this will be used for cross-population based selection analysis
../bin/bcftools view -I -c 2 -C 2 -S yri_ceu_chb_iids.txt -O z 1000genomes_phase3_chr1_150000000_170000000.vcf.gz &gt; yri_ceu_chb_chr1_150000000_170000000_filtered.vcf.gz

# keep individuals only from YRI
# this will be used for haplotype based selection analysis
../bin/bcftools view -S yri_iids.txt -O z yri_ceu_chb_chr1_150000000_170000000_filtered.vcf.gz &gt; yri_chr1_150000000_170000000_filtered.vcf.gz

# write bim file from vcf (-H supress header)
../bin/bcftools view yri_ceu_chb_chr1_150000000_170000000_filtered.vcf.gz -H | awk &#39;{print $1, $3, &quot;0&quot;, $2, $4, $5}&#39; &gt; yri_ceu_chb_chr1_150000000_170000000_filtered.bim

# get genetic position + interpolate variants not present in the genetic_map
../bin/plink --bim yri_ceu_chb_chr1_150000000_170000000_filtered.bim \
             --cm-map genetic_map_chr1_combined_b37.txt 1 \
             --make-just-bim \
             --out yri_ceu_chb_chr1_150000000_170000000_filtered_genetic_map

awk &#39;{print $1, $2, $3, $4}&#39; yri_ceu_chb_chr1_150000000_170000000_filtered_genetic_map.bim &gt; yri_ceu_chb_chr1_150000000_170000000_filtered_genetic_map.map 
</code></pre>

<h3>Questions</h3>

<h2>2. Computing Selection Statistics</h2>

<h3>2.1 Fst</h3>

<pre><code class="bash"># Fst between YRI and CEU
../bin/plink --vcf yri_ceu_chb_chr1_150000000_170000000_filtered.vcf.gz \
             --fst \
             --within yri_ceu_clst.txt \
             --double-id \
             --id-delim . \
             --out yri_ceu_chr1_150000000_170000000_filtered \
             --set-missing-var-ids @:#

# Fst between YRI and CHB
../bin/plink --vcf yri_ceu_chb_chr1_150000000_170000000_filtered.vcf.gz \
             --fst \
             --within yri_chb_clst.txt \
             --double-id \
             --id-delim . \
             --out yri_chb_chr1_150000000_170000000_filtered \
             --set-missing-var-ids @:#

# Fst between CEU and CHB
../bin/plink --vcf yri_ceu_chb_chr1_150000000_170000000_filtered.vcf.gz \
             --fst \
             --within ceu_chb_clst.txt \
             --double-id \
             --id-delim . \
             --out ceu_chb_chr1_150000000_170000000_filtered \
             --set-missing-var-ids @:#
</code></pre>

<h3>2.2. PBS</h3>

<p>Not sure if PBS that useful here unless were intersted in frequency change accumlated in YRI branch since out of Africa?</p>

<p>\[  T = -log(1 - Fst)  \]
\[  PBS = \frac{T_{YRI,CEU} + T_{YRI,CHB} - T_{CEU,CHB}}{2}  \]</p>

<pre><code class="R">library(dplyr)

fst_yri_ceu_df &lt;- read.table(&#39;yri_ceu_chr1_150000000_170000000_filtered.fst&#39;, header = TRUE)
fst_yri_chb_df &lt;- read.table(&#39;yri_chb_chr1_150000000_170000000_filtered.fst&#39;, header = TRUE)
fst_ceu_chb_df &lt;- read.table(&#39;ceu_chb_chr1_150000000_170000000_filtered.fst&#39;, header = TRUE)

# bug - too many nans
pbs_df &lt;- data.frame(CHR = fst_yri_ceu_df$CHR, SNP=fst_yri_ceu_df$SNP, POS=fst_yri_ceu_df$POS,
                     NMISS = fst_yri_ceu_df$NMISS, YRI_CEU_FST = fst_yri_ceu_df$FST, 
                     YRI_CHB_FST = fst_yri_chb_df$FST, CEU_CHB_FST = fst_ceu_chb_df$FST)
</code></pre>

<h3>2.3 iHS</h3>

<pre><code class="bash">../bin/selscan --ihs \
               --vcf yri_ceu_chb_chr1_150000000_170000000_filtered.vcf.gz \
               --map yri_ceu_chb_chr1_150000000_170000000_filtered_genetic_map.map \
               --maf 0.01 \
               --out yri_ceu_chb_chr1_150000000_170000000_filtered   
</code></pre>

<h3>2.4 nSL (maybe)</h3>

<h3>2.4 Visualization</h3>

<h3>Questions</h3>

<h2>3. Assesing Significance</h2>

<h3>3.1 Genome(Region)-Wide Null Distribution</h3>

<h3>3.2 Simulating a Null Model</h3>

<h3>3.3 Visualization</h3>

<h3>3.4 Spandrels of San Marco (Caveats)</h3>

<h3>Questions</h3>

</body>

</html>

