# Understanding Natural Selection from Population Genomic Data

* still in progress (not tested). 
* using EDAR as an example (http://www.snpedia.com/index.php/Rs3827760)
* description / writing needs to be filled in 

## 1. Setting up your Enviroment 

### 1.1 Downloading and Building Methods

#### Download

```bash
cd src/

# download tabix
git clone https://github.com/samtools/tabix

# download htslib & bcftools
wget http://sourceforge.net/projects/samtools/files/samtools/1.2/htslib-1.2.1.tar.bz2
wget http://sourceforge.net/projects/samtools/files/samtools/1.2/bcftools-1.2.tar.bz2

# clone selscan
git clone -b devel https://github.com/szpiech/selscan

# download plink (double check that the correct OS is specified)
wget https://www.cog-genomics.org/static/bin/plink150602/plink_linux_x86_64_dev.zip
```

#### Build

```bash
# tabix
cd tabix/
make
mv tabix ../../bin/

# bcftools
cd ../
tar xvjf htslib-1.2.1.tar.bz2
tar xvjf bcftools-1.2.tar.bz2
cd bcftools-1.2/
make
mv bcftools ../../bin/

# plink
cd ../
mkdir plink
unzip plink_linux_x86_64_dev.zip -d plink
mv plink/plink ../bin

# selscan
cd selscan/

# open Makefile make sure the correct OS is specified 
cd src/ 
make
mv norm ../../../bin/
mv selscan ../../../bin/

# lets go home
cd ../../../
ls bin/

# you should see binaries for tabix, bcftools, plink, selscan, and norm

# export path 
export PATH=path_to_SSW_2015_dir/bin:$PATH
```

### 1.2 Downloading Data

#### 1000genomes_phase3

```bash
cd data/

# download and filter a 3MB region (chr2:108013601-111013601) from the 1000genomes project. Sometimes the ftp goes down if you get an error try running the command again.
tabix -h ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr2.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz 2:108013601-111013601 | bgzip -c > 1000genomes_phase3_chr2_108013601_111013601.vcf.gz

# download 1000genomes_phase3 individual information
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/integrated_call_samples_v3.20130502.ALL.panel
```

### 1.3 Filtering Data and QC

#### Preparing CLST / IID Files 

```bash
# remove the first line and format file as IID | IID | CLST for CHB, YRI, CEU
sed '1d' integrated_call_samples_v3.20130502.ALL.panel | awk '($2 == "CHB" || $2 == "YRI" || $2 == "CEU") {print $1, $1, $2}' > chb_yri_ceu_clst.txt

# write a file of just IIDs 
awk '{print $1}' chb_yri_ceu_clst.txt > chb_yri_ceu_iids.txt

# write a clst file for just CHB and YRI 
awk '($3 == "CHB" || $3 == "YRI") {print $0}' chb_yri_ceu_clst.txt > chb_yri_clst.txt

# write a clst file for just CHB and CEU
awk '($3 == "CHB" || $3 == "CEU") {print $0}' chb_yri_ceu_clst.txt > chb_ceu_clst.txt

# write a clst file for just YRI and CEU
awk  '($3 == "YRI" || $3 == "CEU") {print $0}' chb_yri_ceu_clst.txt > yri_ceu_clst.txt

# write a file of just IIDs from CHB 
awk '($3 == "CHB") {print $1}' chb_yri_ceu_clst.txt > chb_iids.txt

# check if # individuals are correct
wc -l chb_yri_ceu_iids.txt chb_yri_clst.txt chb_ceu_clst.txt yri_ceu_clst.txt chb_iids.txt
```

#### Preparing Genomic Data

```bash
# filter out indels (-V indels), multi-allelic sites (-m 2 -M 2), keep individuals (-S) from YRI, CEU, CHB, and filter out variants with maf below 10%
# this will be used for cross-population based selection analysis
bcftools view -v snps -m 2 -M 2 -S chb_yri_ceu_iids.txt -O z -q '.1 minor' 1000genomes_phase3_chr2_108013601_111013601.vcf.gz > chb_yri_ceu_chr2_108013601_111013601_filtered.vcf.gz

# keep YRI individuals from the filtered vcf
# this will be used for haplotype based selection analysis.
bcftools view -S chb_iids.txt -O z chb_yri_ceu_chr2_108013601_111013601_filtered.vcf.gz > chb_chr2_108013601_111013601_filtered.vcf.gz

# write bim file from vcf (-H supress header)
bcftools view chb_chr2_108013601_111013601_filtered.vcf.gz -H | awk '{print $1, $3, "0", $2, $4, $5}' > chr2_108013601_111013601_filtered.bim

# get genetic positions and interpolate genetic positions for variants not present in the genetic_map
plink --bim chr2_108013601_111013601_filtered.bim \
      --cm-map genetic_map_chr2_combined_b37.txt 2 \
      --make-just-bim \
      --out chr2_108013601_111013601_filtered_gm

# format as map file
awk '{print $1, $2, $3, $4}' chr2_108013601_111013601_filtered_gm.bim > chr2_108013601_111013601_filtered_gm.map 
```

### Questions

## 2. Computing Selection Statistics 

### 2.1 Fst

```bash
# Fst between CHB and YRI
plink --vcf chb_yri_ceu_chr2_108013601_111013601_filtered.vcf.gz \
      --fst \
      --within chb_yri_clst.txt \
      --double-id \
      --id-delim . \
      --out chb_yri_chr2_108013601_111013601_filtered \
      --set-missing-var-ids @:#
     
# Fst between CHB and CEU
plink --vcf chb_yri_ceu_chr2_108013601_111013601_filtered.vcf.gz \
      --fst \
      --within chb_ceu_clst.txt \
      --double-id \
      --id-delim . \
      --out chb_ceu_chr2_108013601_111013601_filtered \
      --set-missing-var-ids @:#
             
# Fst between YRI and CEU
plink --vcf chb_yri_ceu_chr2_108013601_111013601_filtered.vcf.gz \
      --fst \
      --within yri_ceu_clst.txt \
      --double-id \
      --id-delim . \
      --out yri_ceu_chr2_108013601_111013601_filtered \
      --set-missing-var-ids @:#
```

```R
library(dplyr)

# read data
fst_chb_yri_df <- read.table('chb_yri_chr2_108013601_111013601_filtered.fst', header = TRUE)
fst_chb_ceu_df <- read.table('chb_ceu_chr2_108013601_111013601_filtered.fst', header = TRUE)
fst_yri_ceu_df <- read.table('yri_ceu_chr2_108013601_111013601_filtered.fst', header = TRUE)

# join data.frames, filter out rows with missing variants from any population, rename cols and select rows of interest for plotting later
fst_df <- fst_chb_yri_df %>% 
          inner_join(fst_chb_ceu_df, by = c('CHR', 'SNP', 'POS')) %>% 
          inner_join(fst_yri_ceu_df, by = c('CHR', 'SNP', 'POS')) %>%
          filter(is.nan(FST) == FALSE, is.nan(FST.x) == FALSE, is.nan(FST.y) == FALSE) %>%
          rename(FST_CHB_YRI = FST, FST_CHB_CEU = FST.x, FST_YRI_CEU = FST.y) %>%
          select(SNP, POS, FST_CHB_YRI, FST_CHB_CEU, FST_YRI_CEU)
```

### 2.2. PBS

$$ T = -log(1 - Fst) $$
$$ PBS = \frac{T_{CHB,YRI} + T_{CHB,CEU} - T_{YRI,CEU}}{2} $$

```R
# compute pbs and make a data.frame that holds all the information we have computed thus far
cross_pop_df <- fst_df %>% 
                mutate(T_CHB_YRI = -log(1 - FST_CHB_YRI), T_CHB_CEU = -log(1 - FST_CHB_CEU), 
                       T_YRI_CEU = -log(1 - FST_YRI_CEU)) %>%
                mutate(PBS = (T_CHB_YRI + T_CHB_CEU - T_YRI_CEU) / 2) %>%
                select(SNP, POS, FST_CHB_YRI, FST_CHB_CEU, FST_YRI_CEU, PBS)
```

### 2.3 iHS

this region might be too big (i.e. too slow for the workshop) perhaps a more stringent maf filter would be reasonable? 

```bash
# compute iHS! This will take about 10 minutes ...
selscan --ihs \
        --vcf chb_chr2_108013601_111013601_filtered.vcf.gz \
        --map chr2_108013601_111013601_filtered_gm.map \
        --out chb_chr2_108013601_111013601_filtered   
```

### Questions

## 3. Assesing Significance 

### 3.1 Genome(Region)-Wide Null Distribution

#### normalize iHS
```
norm --ihs --files chb_chr2_108013601_111013601_filtered.ihs.out
```

#### normalize PBS 

to be filled

### 3.2 Simulating a Null Model

### 3.3 Visualization 

needs editing still!

```R
library(ggplot2)
library(reshape2)

ihs_df <- read.table("chb_chr2_108013601_111013601_filtered.ihs.out.100bins.norm", header = FALSE)
colnames(ihs_df) <- c("SNP", "POS", "FREQ", "ihh1", "ihh0", "unstandardized_iHS",
                      "standardized_iHS", "top")

selection_df <- ihs_df %>% inner_join(cross_pop_df, by = c("SNP", "POS")) %>% mutate(standardized_iHS = abs(standardized_iHS))
melted_selection_df <- selection_df %>% select(POS, standardized_iHS, PBS, FST_CHB_YRI, FST_CHB_CEU, FST_YRI_CEU) %>% melt(id = c("POS"))

p <- ggplot(melted_selection_df, aes(x = POS, y = value)) + geom_point() + 
     xlab('Position on chr2') + theme_bw() + facet_grid(variable ~ ., scales = "free_y") 
```

### 3.4 Spandrels of San Marco (Caveats)

### Questions 

## 4. References








